src/odemis/driver/focustracker.py

```diff
@@ -39,164 +43,162 @@ def __init__(self, name, role, channel, node_idx, **kwargs):
-        self.proportional_gain_sdo = self.node.sdo['CO Proportional Band Xp1'][1]
-        self.integral_gain_sdo = self.node.sdo['CO Integral Action Time Tn1'][1]
-        self.derivative_gain_sdo = self.node.sdo['CO Derivative Action Time Tv1'][1]
+        self._proportional_gain_sdo = self.node.sdo['CO Proportional Band Xp1'][1]
+        self._integral_gain_sdo = self.node.sdo['CO Integral Action Time Tn1'][1]
+        self._derivative_gain_sdo = self.node.sdo['CO Derivative Action Time Tv1'][1]
 
         self._tracking_sdo = self.node.sdo['Controller On/ Off'][1]
         # set VAs only if there is hardware, channel = '' if no hardware
-        if channel != '':
-            self._metadata[model.MD_GAIN_P] = self.proportional_gain_sdo.raw
-            self._metadata[model.MD_GAIN_I] = self.integral_gain_sdo.raw
-            self._metadata[model.MD_GAIN_I] = self.derivative_gain_sdo.raw
-            self.target_pos = model.FloatVA(self._target_pos_sdo.raw, unit="m", readonly=True,
-                                            getter=self.get_target_pos)
-            self.current_pos = model.FloatVA(self._current_pos_sdo.raw, unit="m", readonly=True,
-                                             getter=self.get_current_pos)
-            self.tracking = model.BooleanVA(self._tracking_sdo.raw, getter=self.get_tracking, setter=self.set_tracking)
+        self._metadata[model.MD_GAIN_P] = self._proportional_gain_sdo.raw
+        self._metadata[model.MD_GAIN_I] = self._integral_gain_sdo.raw
+        self._metadata[model.MD_GAIN_I] = self._derivative_gain_sdo.raw
+        self.targetPosition = model.FloatContinuous(self._target_pos_sdo.raw, TARGET_POSITION_RANGE, unit="m",
+                                                    getter=self._get_target_pos, setter=self._set_target_pos)
+        self.position = model.FloatVA(self._position_sdo.raw, unit="m", readonly=True, getter=self._get_position)
+        self.tracking = model.BooleanVA(self._tracking_sdo.raw, getter=self._get_tracking,
+                                        setter=self._set_tracking)
 
     def terminate(self):
         """Disconnect from CAN bus."""
         if self.network:
+            self._set_tracking(False)
             self.network.sync.stop()
             self.network.disconnect()
             self.network = None
 
     def updateMetadata(self, md):
         if model.MD_GAIN_P in md:
-            md[model.MD_GAIN_P] = self.set_proportional(md[model.MD_GAIN_P])
+            md[model.MD_GAIN_P] = self._set_proportional(md[model.MD_GAIN_P])
         if model.MD_GAIN_I in md:
-            md[model.MD_GAIN_I] = self.set_integral(md[model.MD_GAIN_I])
+            md[model.MD_GAIN_I] = self._set_integral(md[model.MD_GAIN_I])
         if model.MD_GAIN_D in md:
-            md[model.MD_GAIN_D] = self.set_derivative(md[model.MD_GAIN_D])
+            md[model.MD_GAIN_D] = self._set_derivative(md[model.MD_GAIN_D])
         super(FocusTrackerCO, self).updateMetadata(md)
```

Comment:
A little annoying corner case is when the user request a new set of PID values, and one of the value is rejected. For example, let's say that D is a negative value. Then, P and I will be set, then updateMetadata will raise an error. The P&I will be updated on the device, but the metadata left as-is.

Maybe the simplest way to deal with it would be to immediately update the metadata after setting each gain.